titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/DefaultKubeJobManagementReconciler.java

```diff
@@ -0,0 +1,286 @@
+
+        Map<OrphanedKind, List<Task>> orphanedTasksByKind = new HashMap<>();
+        for (Task task : tasksNotInApiServer) {
+            if (task.getStatus().getState().equals(TaskState.KillInitiated)) {
+                orphanedTasksByKind.computeIfAbsent(OrphanedKind.KILL_INITIATED, s -> new ArrayList<>()).add(task);
+            } else {
+                if (findNode(task, nodes).isPresent()) {
+                    orphanedTasksByKind.computeIfAbsent(OrphanedKind.UNKNOWN, s -> new ArrayList<>()).add(task);
+                } else {
+                    orphanedTasksByKind.computeIfAbsent(OrphanedKind.NODE_TERMINATED, s -> new ArrayList<>()).add(task);
+                }
+            }
+        }
+
+        orphanedTasksByKind.forEach((kind, tasks) -> {
+            logger.info("Attempting to transition {} orphaned tasks to finished ({}): {}", tasks.size(), kind, tasks);
+            orphanedTaskGauges.get(kind).set(tasks.size());
+
+            for (Task task : tasks) {
+
+                String reasonCode;
+                String reasonMessage;
+                switch (kind) {
+                    case KILL_INITIATED:
+                        reasonCode = TaskStatus.REASON_TASK_KILLED;
+                        reasonMessage = "Task killed";
+                        break;
+                    case NODE_TERMINATED:
+                        reasonCode = TaskStatus.REASON_TASK_LOST;
+                        reasonMessage = "Agent terminated";
+                        break;
+                    case UNKNOWN:
+                    default:
+                        reasonCode = TaskStatus.REASON_TASK_LOST;
+                        reasonMessage = "Task lost between control plane and machine";
+                        break;
+                }
+
+                publishEvent(task,
+                        TaskStatus.newBuilder()
+                                .withState(TaskState.Finished)
+                                .withReasonCode(reasonCode)
+                                .withReasonMessage(reasonMessage)
+                                .withTimestamp(clock.wallTime())
+                                .build()
+                );
+            }
+            logger.info("Finished orphaned task transitions to finished ({})", kind);
+        });
+    }
+
+    private Optional<V1Node> findNode(Task task, Map<String, V1Node> nodes) {
+        String instanceId = task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_AGENT_INSTANCE_ID);
+        if (instanceId == null) {
+            return Optional.empty();
+        }
+        return Optional.ofNullable(nodes.get(instanceId));
+    }
+
+    private boolean shouldTaskBeInApiServer(Task task) {
+        boolean isRunning = TaskState.isRunning(task.getStatus().getState());
+
+        if (JobFunctions.isOwnedByKubeScheduler(task)) {
+            return isRunning || (task.getStatus().getState() == TaskState.Accepted && TaskStatus.hasPod(task));
```

Comment:
In the `TaskState.Accepted && hasPod()` case, I think we need a grace time to allow state propagation through kube-apiserver and our informers, otherwise we may be killing tasks prematurely